### ISSUE TITLE
Opt-in

### ISSUE THREAD

**Issue by bakkot** [jasnell/proposal-construct#1]
I'd kind of prefer this to be opt-in, mostly to address @ljharb's concern about installing internal slots (or private fields) on arbitrary objects.

For example, a new contextual `callable` keyword at class declaration time:

```js
callable class A {
  constructor() {
    this.x = 0;
  }
}

function B () {
  A.call(this);
}

(new B).x; // 0
```

**Comment by ljharb** [comment]
Indeed, it would have to be opt-in inside the `class` syntax to be safe.

This seems like @allenwb's callable classes proposal(s), however.

**Comment by jasnell** [comment]
the opt in would occur in B ... which would essentially serve as a go-between on the regular class and any old style subclasses... e.g.

```js
class A {
  constructor(a) {
    this.a =a ;
  }
}

function B(a) {
  A.construct(this, [a]);
}

function C(a) {
  // Old style
  B.call(this, a);
}

var c = new C();
```

For Node.js core, we would essentially be able to define our implementation using classes, but expose those using the pattern in B, so that downstream code written like C would just still continue to work.

**Comment by ljharb** [comment]
right, but that's the problem - that something from outside of `A` can get access to its internal slot structure without using `extends`.

**Comment by jasnell** [comment]
We would have the exact same problem by making class constructors callable by default. Perhaps a new kind of class then? A Callable Class....

```js
class* Foo {}

Foo(); // still throws because there's no this 
new Foo(); // constructs like a normal class
Foo.call(this); // bingo! It works
Foo.bind(this)(); // this works too since there's a this
```

I dunno, just brainstorming at this point

**Comment by ljharb** [comment]
In your a b c examples, I'd think making B a `callable class extends a` might be the best solution.

**Comment by bakkot** [comment]
(Incidentally, if we get class decorators, I'd definitely want any `callable` annotation to be a decorator feature, rather than its own keyword.)

**Comment by jasnell** [comment]
Yeah, stewing on that. Thinking through any potential issues having a callable class extend a normal class.

**Comment by bakkot** [comment]
Oh, yeah. I guess you'd want to require that any superclass of a callable class itself be callable, otherwise it's not really opt-in. (Though, `super()` is normally the thing which *defines* `this`, so it's kind of weird either way.)
